hw_strings.py

```diff
@@ -0,0 +1,178 @@
+    group_keys = []
+    for i in fdata:
+        if i[grouper] not in group_keys:
+            group_keys.append(i[grouper])
+    grouped_dict = {}
+    for x in group_keys:
+        temp_values_list = []
+        for j in fdata:
+            if j[grouper] == x:
+                t = dict(j)
+                del t[grouper]
+                temp_values_list.append(t)
+        grouped_dict[x] = temp_values_list
+    return grouped_dict
+
+
+def get_most_popular_period(periods_list):
+    """Function accepts list of time periods(hours, days, etc.). Return value is a string that contains 1 period in case
+    there is only one most frequent period in periods_list
+    or several comma-separated periods with the same frequency as max"""
+    periods = dict(Counter(periods_list))
+    counter_period = Counter(periods_list).most_common(1)[0][0]
+    pop_period = [counter_period]
+    for key in periods.keys():
+        if key != counter_period and periods[key] == periods[counter_period]:
+            pop_period.append(key)
+    most_popular_period = ', '.join(p for p in pop_period)
+    return most_popular_period
+
+
+def get_info_by_ip(src, target):
+    """Function takes .txt file with info on ip address, visiting time and week day  according to following pattern:
+     139.18.150.126 23:12:44 sunday
+     Output: created target file with following information:
+     unique ip-addresses from src file, number of visits, the most popular day of the week and at the end the most
+     popular visiting hour(s) is specified for the whole site (all ip addresses).
+     Example:
+     IP Address   Visits Day
+     139.18.150.126  9   sunday
+     The most popular hour(s): 11:00
+     """
+    with open(src, encoding='utf-8') as src_file:
+        # Saving data to ip_data list of dicts
+        ip_data = []
+        result = []
+        for line in src_file:
+            line = line.split()
+            info_dict = {"ip": line[0], "time": line[1], "day": line[2]}
+            ip_data.append(info_dict)
+        # my_printer = pprint.PrettyPrinter(width=90, sort_dicts=False)
+        # my_printer.pprint(ip_data)
+        ip_groups = group_data_by(ip_data, 'ip')
+        # my_printer.pprint(ip_groups)
+        unique_ip_addr = list(ip_groups.keys())
+        # Collecting info on visits and days per IP address
+        for ip in unique_ip_addr:
+            visits = len(ip_groups[ip])
+            # Collecting info on days per ip address
+            days = []
+            for item in ip_groups[ip]:
+                days.append(item['day'])
+            popular_day = get_most_popular_period(days)
+            output_dict = {"IP Address": ip, "Visits": visits, "Day": popular_day}
+            result.append(output_dict)
+
+        # Collecting visits hours data
+        visit_hours = []
+        for data in ip_data:
+            visit_hours.append(data['time'].split(':')[0] + ':00')
+        popular_hour = get_most_popular_period(visit_hours)
```

Comment:
`with open` block try to keep as small as you can. The file should be closed after getting info as fast as we can.